### PR TITLE
Config FollowUp

### PR DESCRIPTION
**Pull request by nupplaphil**
FollowUp #6581 

This is based on the suggestion of https://github.com/friendica/friendica/pull/6581#discussion_r253332518

#### New Instances
1. New `Configuration` instance (business logic for using Cache and/or Adapter)
2. New `PConfiguration` instance (business logic for using Cache and/or Adapter)
3. New `DBFactory::init()` for `DBA::connect()` in a more modularized way

### Refactoring 
1. Decoupled `JIT-/PreloadConfigAdapter` from `ConfigCache` (it is now more like a service-hull for DBA-calls)
2. Decoupled `JIT-/PreloadPConfigAdapter` from `ConfigCache` (it is now more like a service-hull for DBA-calls)
3. Refactored `ConfigFactory`
4. New method `ConfigCache::has(P)()` method for Cache-checks
5. Refactored `LoggerFactory` (now using directly `Configuration` instead of implicit loading and doesn't need `App` anymore)

### Testing
1. Added `ConfigurationTest`
2. Added `PConfigurationTest`

The overall goal is to have a straight dependency line:
1. `ConfigCache`
2. `DBA::connected()`
3. `(P)Configuration` as overall Configuration instance (should be used at any other places except `DBA`)
4. `Logger`
5. `App`

The main benefit of this PR is that we do have just **one** cache for all the configs instead of two and a more straight logic for using Config-values (I tried to make small and easy readable classes instead of the if-elseif-elseif-else logic). 

I made a lot of tests around the `ConfigCache` and the new `(P)Configuration` instances to make the refactoring as bug-free as possible. Of course, I did some click-tests too.

My node https://friendica.philipp.info is currently running with this branch and I'm watching the logs ..
I think I will let it run about 2-3 days to make sure nothing "bad" happen..
[Edit] I need a little bit of time because I accidentally deleted my `/storage/` directory -> see https://github.com/friendica/docker/pull/51

### CODE REVIEW

**Comment by nupplaphil** [comment]
btw. first of all I'm sorry @MrPetovan .. I'm sure you'll find a lot of improvements here :-/

File: src/Core/Config/Adapter/IConfigAdapter.php [hunk 1]
```diff
@@ -13,52 +13,59 @@ interface IConfigAdapter
...
 
 	/**
 	 * Checks, if the current adapter is connected to the backend
 	 *
 	 * @return bool
 	 */
 	public function isConnected();
+
+	/**
+	 * Checks, if a config value ($value) in the category ($cat) is already loaded.
```
Comment (MrPetovan) [inline src/Core/Config/Adapter/IConfigAdapter.php]: It should be `config key ($key)` instead.

File: src/Core/Config/Adapter/IPConfigAdapter.php [hunk 2]
```diff
@@ -20,7 +20,7 @@ interface IPConfigAdapter
 	 * @param string $uid The user_id
 	 * @param string $cat The category of the configuration value
 	 *
-	 * @return void
+	 * @return array
```
Comment (MrPetovan) [inline src/Core/Config/Adapter/IPConfigAdapter.php]: Please update the description of the method if it now returns something.

File: src/Core/Config/Adapter/IConfigAdapter.php [hunk 3]
```diff
@@ -13,52 +13,59 @@ interface IConfigAdapter
 	 *
 	 * @param string  $cat The category of the configuration values to load
 	 *
-	 * @return void
+	 * @return array
```
Comment (MrPetovan) [inline src/Core/Config/Adapter/IConfigAdapter.php]: Please update the description of the method if it now returns something.

File: src/Core/Logger.php [hunk 4]
```diff
@@ -67,73 +65,22 @@ class Logger extends BaseObject
...
 	/**
 	 * Sets the default logging handler for Friendica.
-	 * @todo Can be combined with other handlers too if necessary, could be configurable.
 	 *
 	 * @param LoggerInterface $logger The Logger instance of this Application
-	 *
-	 * @throws InternalServerErrorException if the logger factory is incompatible to this logger
 	 */
-	public static function setLogger($logger)
+	public static function init($logger)
```
Comment (MrPetovan) [inline src/Core/Logger.php]: Please add the `LoggerInterface` type-hint.
```suggestion
	public static function init(LoggerInterface $logger)
```

File: src/Core/Logger.php [hunk 5]
```diff
@@ -67,73 +65,22 @@ class Logger extends BaseObject
...
-	 * @see https://github.com/php-fig/fig-standards/blob/master/accepted/PSR-3-logger-interface.md#5-psrlogloglevel
-	 *
-	 * @param string $level the level to be mapped
+	 * Sets the default dev-logging handler for Friendica.
 	 *
-	 * @return string the PSR-3 compliant level
+	 * @param LoggerInterface $logger The Logger instance of this Application
 	 */
-	private static function mapLegacyConfigDebugLevel($level)
+	public static function setDevLogger($logger)
```
Comment (MrPetovan) [inline src/Core/Logger.php]: Please add the `LoggerInterface` type-hint.
```suggestion
	public static function setDevLogger(LoggerInterface $logger)
```

File: src/Database/DBA.php [hunk 6]
```diff
@@ -582,7 +583,8 @@ public static function p($sql) {
 			self::$errorno = $errorno;
 		}
 
-		$a->saveTimestamp($stamp1, 'database');
+		/// @todo profiling out of App
+		//$a->saveTimestamp($stamp1, 'database');
```
Comment (MrPetovan) [inline src/Database/DBA.php]: Is this PR effectively killing any DB profiling?

File: bin/auth_ejabberd.php [hunk 7]
```diff
@@ -55,11 +55,15 @@
...
-$configLoader = new Config\ConfigCacheLoader($basedir);
-$config = Factory\ConfigFactory::createCache($configLoader);
-$logger = Factory\LoggerFactory::create('auth_ejabberd', $config);
+$configLoader = new Cache\ConfigCacheLoader($basedir);
+$configCache = Factory\ConfigFactory::createCache($configLoader);
+Factory\DBFactory::init($configCache, $_SERVER);
+$config = Factory\ConfigFactory::createConfig($configCache);
+// needed to call PConfig::init()
+Factory\ConfigFactory::createPConfig($configCache);
+Factory\LoggerFactory::create('auth_ejabberd', $config);
```
Comment (MrPetovan) [inline bin/auth_ejabberd.php]: Please move this whole sequence into a static factory method. This is getting way more than you can safely duplicate in each endpoint.

File: src/Core/Config/Adapter/IConfigAdapter.php [hunk 8]
```diff
@@ -13,52 +13,59 @@ interface IConfigAdapter
...
-	 * @param string  $k             The configuration key to query
-	 * @param mixed   $default_value optional, The value to return if key is not set (default: null)
-	 * @param boolean $refresh       optional, If true the config is loaded from the db and not from the cache (default: false)
+	 * @param string  $cat The category of the configuration value
+	 * @param string  $key The configuration key to query
 	 *
-	 * @return mixed Stored value or null if it does not exist
+	 * @return mixed Stored value or "!<unset>!" if it does not exist
 	 */
-	public function get($cat, $k, $default_value = null, $refresh = false);
```
Comment (MrPetovan) [inline src/Core/Config/Adapter/IConfigAdapter.php]: Why did you remove the default value here? Are default values still handled at all?

File: bin/auth_ejabberd.php (hunk 7 above)
Comment (nupplaphil) [reply bin/auth_ejabberd.php]: There's already a feature branch where I use the DI-library exactly for this sequence, so yes of course, I'll move it temporary in a static method :)

File: src/Core/Config/Adapter/IConfigAdapter.php (hunk 8 above)
Comment (nupplaphil) [reply src/Core/Config/Adapter/IConfigAdapter.php]: The `Adapter` and `ConfigCache` aren't responsible for default values anymore. The adapter is for `DBA` backend loadings (JIT or complete) and the `ConfigCache` is responsible for caching config values. The `Configuration` class is the glue between them and is the facade for the outer world (which contains the default  values).

My overall target was to clean up the classes to be small, readable and testable.

File: src/Database/DBA.php (hunk 6 above)
Comment (nupplaphil) [reply src/Database/DBA.php]: I'm afraid.. Yes for now.. There's a loopback to the `App` class for profiling.

I started creating a `Profiler` class to encapsulate the `performance[]` and `callstack[]` information, but the PR is big enough.

If it's a showstopper, I can create a PR for the `Profiler` first.

@ananndo I think this is something important for you too.

File: src/Core/Config/Adapter/IConfigAdapter.php (hunk 8 above)
Comment (nupplaphil) [reply src/Core/Config/Adapter/IConfigAdapter.php]: f.e. (just an idea) we now can easily create a `CacheConfigAdapter` which initially loads the whole config into a key-value storage (f.e. redis) and sync the result in a specific intervall back to the db. It has no impact on the usage of the Config, just another adapter :)

Btw. this would maybe something to rapidly decrease the db-calls for config keys

**Comment by nupplaphil** [comment]
@ all there's a strange issue with `DBA::lastInsertId()`.

The `DBA::insert()` returns true for inserting, but the lastinsertid is 0 every time.. That's the reason for all the failed tests ..

File: src/Core/Config/Configuration.php [hunk 9]
```diff
@@ -0,0 +1,153 @@
...
+	 * @param string  $key           The configuration key to query
+	 * @param mixed   $default_value optional, The value to return if key is not set (default: null)
+	 * @param boolean $refresh       optional, If true the config is loaded from the db and not from the cache (default: false)
+	 *
+	 * @return mixed Stored value or null if it does not exist
+	 */
+	public function get($cat, $key, $default_value = null, $refresh = false)
+	{
+		// if the value isn't loaded or refresh is needed, load it to the cache
+		if ($this->configAdapter->isConnected() &&
```
Comment (nupplaphil) [inline src/Core/Config/Configuration.php]: @MrPetovan This one is the new "magic"

In other words (without the methods around it):
```php
if ($adapter->isConnected && ($adapter->in_db[$cat][$key] || $refresh) {}
```
This is will become the only check for each `Config::get()`, otherwise the Cache will be used. I hope that this statement is fast enough (or at least as fast as before).

File: index.php [hunk 10]
```diff
@@ -4,24 +4,16 @@
...
 
 require __DIR__ . '/vendor/autoload.php';
 
-$basedir = BasePath::create(__DIR__, $_SERVER);
-$configLoader = new Config\ConfigCacheLoader($basedir);
-$config = Factory\ConfigFactory::createCache($configLoader);
-$logger = Factory\LoggerFactory::create('index', $config);
-
 // We assume that the index.php is called by a frontend process
 // The value is set to "true" by default in App
```
Comment (MrPetovan) [inline index.php]: This comment is now invalid.

**Comment by nupplaphil** [comment]
I based this changes on #6673 and force-pushed them again.

File: src/Database/DBA.php (hunk 6 above)
Comment (nupplaphil) [reply src/Database/DBA.php]: will be fixed per #6673

**Comment by annando** [comment]
I have to ask: Had you had you got running these changes for some days running on your production engine? For me the changes are much to massive for being able to check them visually for correctness. So I really prefer having it running on a production system for some days and constantly checking the PHP error logs to see if everything works well.

Same I would like to see for the PR https://github.com/friendica/friendica/pull/6673 and https://github.com/friendica/friendica/pull/6668.

**Comment by nupplaphil** [comment]
The other two were running each 1 day, this one here not because I have to wait for the other two first.

The other two PRs are not that massive as this PR here I guess.
- #6673 introduces a new class and are mainly replacing the calls to this class
- #6668 is introducing a WorkerLogger and are mainly replacing the logger calls (OK and the instancing of the Worker as well)

The current PR on the other hand replaces a lot of Config business logic, so I'm with you, after the merge of the other two, I'd let it run on my node for about 2-3 days

**Comment by nupplaphil** [comment]
@mrpetovan this was a little bit quick .. I wanted to let it run for some days after the merge of #6673 , because this PR based on it.. I hope it's working :-/ ..

**Comment by nupplaphil** [comment]
Addon per https://github.com/friendica/friendica-addons/pull/806

**Comment by MrPetovan** [comment]
> this was a little bit quick .. I wanted to let it run for some days after the merge of #6673 , because this PR based on it.. I hope it's working :-/ ..

In this case, please close your PR while you test on your node. I can't guess if it's ready or not, I just know you submitted a PR and that I can merge it at any time, so I assume it's ready. GitHub also introduced Draft PRs if you want to create a PR but don't want it to be able to be merged just yet.
